lib/bamboo/adapters/smtp_adapter.ex

```diff
@@ -161,6 +168,45 @@ defmodule Bamboo.SMTPAdapter do
     |> add_smtp_line(text_body)
   end
 
+  defp add_multipart_mixed_split(body, delimiter), do: "#{body}\r\n--#{delimiter}\r\n"
+
+  defp add_attachment_header(body, attachment) do
+    << random :: size(32) >> = :crypto.strong_rand_bytes(4)
+    "#{body}Content-Type: attachment.content_type; name=\"#{attachment.filename}\"\r\n" <>
+    "Content-Disposition: attachment; filename=\"#{attachment.filename}\"\r\n" <>
+    "Content-Transfer-Encoding: base64\r\n" <>
+    "X-Attachment-Id: #{random}\r\n"
+  end
+
+  defp add_attachment_body(body, attachment) do
+    data = case attachment.path do
+      {:binary, data} -> data
+      _ -> File.read!(attachment.path)
+    end
+    "#{body}\r\n#{Base.encode64(data)}\r\n"
+  end
+
+  defp add_attachment(attachment, multi_part_mixed_delimiter) do
+    if attachment == nil do
+      ""
+    else
+      ""
+      |> add_multipart_mixed_split(multi_part_mixed_delimiter)
+      |> add_attachment_header(attachment)
+      |> add_attachment_body(attachment)
+    end
+  end
+
+  defp add_attachments(body, %Bamboo.Email{attachments: attachments}, multi_part_mixed_delimiter) do
```

Comment:
What do you think of using pattern matching on function definition?

Something like: 

```elixir
defp add_attachments(body, %Bamboo.Email{attachments: nil}, multi_part_mixed_delimiter), do: body
defp add_attachments(body, %Bamboo.Email{attachments: attachments}, multi_part_mixed_delimiter), do: body
  # ...
end
```